core/domain/exp_services.py

```diff
@@ -2063,6 +2063,345 @@ def get_interaction_id_for_state(exp_id, state_name):
+            )
+    )
+
+    # If the most recently reached checkpoint doesn't exist in current
+    # exploration.
+    if (
+        most_recently_reached_checkpoint_in_current_exploration !=
+        model.most_recently_reached_checkpoint_state_name
+    ):
+        model.most_recently_reached_checkpoint_state_name = (
+            most_recently_reached_checkpoint_in_current_exploration)
+        model.most_recently_reached_checkpoint_exp_version = (
+            latest_exploration.version)
+        model.update_timestamps()
+        model.put()
+
+    # If the furthest reached checkpoint doesn't exist in current
+    # exploration.
+    if (
+        furthest_reached_checkpoint_in_current_exploration !=
+        model.furthest_reached_checkpoint_state_name
+    ):
+        model.furthest_reached_checkpoint_state_name = (
+            furthest_reached_checkpoint_in_current_exploration)
+        model.furthest_reached_checkpoint_exp_version = (
+            latest_exploration.version)
+        model.update_timestamps()
+        model.put()
+
+    return exp_fetchers.get_logged_out_user_progress(unique_progress_url_id)
+
+
+def sync_logged_out_learner_progress_with_logged_in_progress(
+    user_id, unique_progress_url_id):
+
+    """Syncs logged out and logged in learner progress."""
+
+    logged_out_user_data = (
+        exp_fetchers.get_logged_out_user_progress(unique_progress_url_id))
+
+    exp_id = logged_out_user_data.exploration_id
+    latest_exploration = exp_fetchers.get_exploration_by_id(exp_id)
+    exp_user_data = exp_fetchers.get_exploration_user_data(
+        user_id,
+        exp_id
+    )
+
+    logged_in_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exp_id)
+
+    if logged_in_user_model is None:
+        logged_in_user_model = user_models.ExplorationUserDataModel.create(
+            user_id, exp_id)
+
+        logged_in_user_model.most_recently_reached_checkpoint_exp_version = (
+            logged_out_user_data.most_recently_reached_checkpoint_exp_version
+        )
+        logged_in_user_model.most_recently_reached_checkpoint_state_name = (
+            logged_out_user_data.most_recently_reached_checkpoint_state_name
+        )
+        logged_in_user_model.furthest_reached_checkpoint_exp_version = (
+            logged_out_user_data.furthest_reached_checkpoint_exp_version
+        )
+        logged_in_user_model.furthest_reached_checkpoint_state_name = (
+            logged_out_user_data.furthest_reached_checkpoint_state_name
+        )
+        logged_in_user_model.update_timestamps()
+        logged_in_user_model.put()
+
+    elif logged_in_user_model.most_recently_reached_checkpoint_exp_version == logged_out_user_data.most_recently_reached_checkpoint_exp_version: # pylint: disable=line-too-long
+        most_recently_reached_checkpoint_index_in_logged_in_progress = (
+            user_services.get_checkpoints_in_order(
+                latest_exploration.init_state_name,
```

Comment:
Is it true that if logged in and logged out user models have the same exp versions, it will always be the latest exploration exp version? I don't think so. Can u update this with either logged in or logged out user models

Reply:
Fixed.